File: python/zkrypto/chaum_pedersen.py
import random
import hashlib

from zkrypto.prime import Prime
from zkrypto.generator import Generator

class ChaumPedersenProver:

    def __init__(self, bits=20):
        # Generate parameters (p, q, g) using Generator module
        p, q, G = Generator.get_generator_prime(random.randint(bits >> 1, bits), 2)
        self.p = p
        self.q = q
        self.g = G[0]
        self.h = G[1]
        self.k_bl = set() # k blacklist to avoid repeating

    def params(self):
        # Returns the parameters
        return (self.p, self.q, self.g, self.h)

    
    def gen_keypair(self):
        # Generate a keypair (private key x, public key y) given p, g and h
        # Generate private key x
        self._x = random.randint(1, self.p - 2)
        # Compute public key y = g^x mod p
        self.y1 = pow(self.g, self._x, self.p)
        self.y2 = pow(self.h, self._x, self.p)
        return self._x, self.y1, self.y2

    def prove(self):
        # Prover's part of the Schnorr protocol

        # Generate key pair
        self.gen_keypair()

        # Choose a random nonce k
        ## Problem: You need to keep track of which k you use.
        ## If you repeat the k, you leak your x
        k = random.randint(1, self.p - 2)
        while k in self.k_bl:
            k = random.randint(1, self.p - 2)
        self.k_bl.add(k)
        
        # Compute commitment t = g^k mod p
        r1 = pow(self.g, k, self.p)
        r2 = pow(self.h, k, self.p)

        # Compute challenge c = H(r1 || r2 || y1 || y2)
        c = int(hashlib.sha256((str(r1) + str(r2) + str(self.y1) + str(self.y2)).encode()).hexdigest(), 16)
        
        # Compute response s = (k + x * e) % q
        s = (k - (self._x * c % self.q)) % self.q
        return self.y1, self.y2, r1, r2, c, s


class ChaumPedersenVerifier():
    def __init__(self, p, q, g, h):
        self.p = p
        self.q = q
        self.g = g
        self.h = h
    
    
    def verify(self, y1, y2, r1, r2, c, s):
        # Verifier's part of the Schnorr protocol
        # Compute r' = (g^s * y^(-e)) % p
        r1_prime = (pow(self.g, s, self.p) * pow(y1, c, self.p)) % self.p
        r2_prime = (pow(self.h, s, self.p) * pow(y2, c, self.p)) % self.p
        return r1 == r1_prime and r2 == r2_prime

File: src/zkrypto/chaum_pedersen.rs
use num_bigint::{BigInt, RandBigInt};
use num_traits::{Zero, One};
use log::{debug};
use std::fmt::Debug;

use crate::zkrypto::generator::{Generator, get_default};

// Represents a key pair containing private and public components
#[derive(Default)]
pub struct KeyPair{
    x: BigInt,      // Private key
    pub y1: BigInt, // Public key component 1
    pub y2: BigInt  // Public key component 2
}

impl KeyPair {
    // Generates a new key pair using provided generator and parameters
    fn new(g: &BigInt, h: &BigInt, p: &BigInt) -> KeyPair {
        let mut rng = rand::thread_rng();
        let x = rng.gen_bigint_range(&BigInt::one(), p);
        let y1 = g.modpow(&x, p);
        let y2 = h.modpow(&x, p);
        return KeyPair {
            x: x,
            y1: y1,
            y2: y2
        };
    }
}

// Represents a prover for the Chaum-Pedersen protocol
#[derive(Default)]
pub struct ChaumPedersenProver {
    g: Generator,      // Generator instance
    pub kp: KeyPair,   // Key pair
    k: BigInt,         // Random integer k
    r1: BigInt,        // First response value
    r2: BigInt,        // Second response value
    c: BigInt,         // Challenge value
    s: BigInt,         // Solution value
}

impl Debug for ChaumPedersenProver {
    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
        write!(f, "ChaumPedersenProver <-")
    }
}

impl ChaumPedersenProver {
    // Creates a new instance of ChaumPedersenProver
    pub fn new() -> ChaumPedersenProver {
        let g = get_default();
        let kp = KeyPair::new(&g.g, &g.h, &g.p);

        return ChaumPedersenProver {
            g: g,
            kp: kp,
            k: BigInt::zero(),
            r1: BigInt::zero(),
            r2: BigInt::zero(),
            c: BigInt::zero(),
            s: BigInt::zero(),
        }
    }

    // Proves part A of the protocol synchronously
    pub fn prove_sync_a(&mut self) -> (BigInt, BigInt) {
        debug!("[P] PROVE SYNC A");
        let mut rng = rand::thread_rng();

        self.k =  rng.gen_bigint_range(&BigInt::one(), &(self.g.p.clone() - 2));
        self.r1 = self.g.g.clone().modpow(&self.k, &self.g.p);
        self.r2 = self.g.h.clone().modpow(&self.k, &self.g.p);

        debug!("[P] y1: {}, y2:{}", &self.kp.y1, &self.kp.y2);
        debug!("[P] r1: {}, r2:{}", &self.r1, &self.r2);
        debug!("[P] END PROVE SYNC A");
        return (self.r1.clone(), self.r2.clone());
    }

    // Proves part B of the protocol synchronously
    pub fn prove_sync_b(&mut self, c: &BigInt) -> BigInt {
        debug!("[P] PROVE SYNC B");
        self.c = c.clone();
        self.s = (self.k.clone() - ((self.kp.x.clone() * self.c.clone()) % self.g.q.clone())).modpow(&BigInt::one(), &self.g.q);
        debug!("[P] c: {}, s:{}", &self.c, &self.s);
        debug!("[P] END PROVE SYNC B");
        return self.s.clone();
    }
}

// Represents a verifier for the Chaum-Pedersen protocol
#[derive(Default)]
pub struct ChaumPedersenVerifier {
    g : Generator,      // Generator instance
    y1: BigInt,         // Public key component 1
    y2: BigInt,         // Public key component 2
    r1: BigInt,         // First response value
    r2: BigInt,         // Second response value
    c: BigInt,          // Challenge value
    s: BigInt,          // Solution value
}

impl Debug for ChaumPedersenVerifier {
    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
        write!(f, "ChaumPedersenVerifier <-")
    }
}

impl ChaumPedersenVerifier {
    // Creates a new instance of ChaumPedersenVerifier
    pub fn new(y1: BigInt, y2: BigInt) -> ChaumPedersenVerifier {
        let g = get_default();
        return ChaumPedersenVerifier {
            g: g,
            y1: y1.clone(),
            y2: y2.clone(),
            r1: BigInt::zero(),
            r2: BigInt::zero(),
            c: BigInt::zero(),
            s: BigInt::zero(),
        }
    }

    // Verifies part A of the protocol synchronously
    pub fn verify_sync_a(&mut self, r1: BigInt, r2: BigInt) -> BigInt {
        debug!("[V] VERIFY SYNC A");
        let mut rng = rand::thread_rng();

        self.c = rng.gen_bigint_range(&BigInt::one(), &(self.g.p.clone()));
        self.r1 = r1;
        self.r2 = r2;
        debug!("[V] y1: {}, y2:{}", &self.y1, &self.y2);
        debug!("[V] r1: {}, r2:{}", &self.r1, &self.r2);
        debug!("[V] END VERIFY SYNC A");
        return self.c.clone();
    }

    // Verifies part B of the protocol synchronously
    pub fn verify_sync_b(&mut self, s: BigInt) -> bool {
        debug!("[V] VERIFY SYNC B");
        self.s = s;
        let r1_prime: BigInt = (
            self.g.g.modpow(&self.s, &self.g.p) * 
            self.y1.modpow(&self.c, &self.g.p)
        ) % self.g.p.clone();
        let r2_prime: BigInt = (
            self.g.h.modpow(&self.s, &self.g.p) * 
            self.y2.modpow(&self.c, &self.g.p)
        ) % self.g.p.clone();

        debug!("[V] END VERIFY SYNC B");
        return r1_prime == self.r1 && r2_prime == self.r2;
    }
}


#[cfg(test)]
mod test{
    use crate::zkrypto::generator::get_default;
    use crate::zkrypto::chaum_pedersen::{KeyPair, ChaumPedersenProver, ChaumPedersenVerifier};

    #[test]
    fn test_keypair_creation() {
        let g = get_default().g;
        let h = get_default().h;
        let p = get_default().p;

        let keypair = KeyPair::new(&g, &h, &p);

        let y1_expected = g.modpow(&keypair.x, &p);
        let y2_expected = h.modpow(&keypair.x, &p);

        assert_eq!(keypair.y1, y1_expected);
        assert_eq!(keypair.y2, y2_expected);
    }

    #[test]
    fn test_chaum_pedersen_protocol() {
        let mut prover = ChaumPedersenProver::new();
        let mut verifier = ChaumPedersenVerifier::new(prover.kp.y1.clone(), prover.kp.y2.clone());

        // Prover's side
        let (r1, r2) = prover.prove_sync_a();
        let c = verifier.verify_sync_a(r1.clone(), r2.clone());
        let s = prover.prove_sync_b(&c);
        let result = verifier.verify_sync_b(s);

        assert_eq!(result, true);
    }
}